### PR TITLE
KeyStore implementation + key derivation

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

File: polkadot/keystore/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		}
+	}
+
+	fn decrypt(&self, password: &str) -> Result<[u8; PKCS_LEN]> {
+		let (derived_left_bits, derived_right_bits) =
+			crypto::derive_key_iterations(password, &self.salt, self.iterations);
+
+		let mac = crypto::derive_mac(&derived_right_bits, &self.ciphertext).keccak256();
+
+		if mac != self.mac {
```
Comment (NikVolf) [inline polkadot/keystore/src/lib.rs]: might be timing attack vector (was a topic in the recent audit)
though i failed to see how to exploit this in parity setup, better to use constant time comparison to be sure (https://github.com/paritytech/parity/pull/8113)

in general, non-constant comparison breaks mac semantic security

Comment (arkpar) [reply polkadot/keystore/src/lib.rs]: `ring::constant_time` can be used for that

File: polkadot/keystore/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		buf
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use tempdir::TempDir;
+
+    #[test]
```
Comment (gavofyork) [inline polkadot/keystore/src/lib.rs]: spaces. you toying with me here? :-P

File: polkadot/keystore/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+use std::path::PathBuf;
+use std::fs::{self, File};
+use std::io::{self, Write};
+
+use crypto::Keccak256;
+use ed25519::{Pair, Public, PKCS_LEN};
+
+pub use crypto::KEY_ITERATIONS;
+
+error_chain! {
```
Comment (gavofyork) [inline polkadot/keystore/src/lib.rs]: don't these usually go in their own module?

**Pull request merged**